### PR TITLE
Add Logistic Regression algorithm

### PR DESCRIPTION
**Pull request by Zhangxunmt**
Signed-off-by: Xun Zhang <xunzh@amazon.com>

### Description
Add Tribuo Logistic Regression into ML-Common.
UTs and ITs of logistic regression are included in the PR. Addressed comments from the previous PR.

### Issues Resolved
https://github.com/opensearch-project/ml-commons/issues/318
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/TribuoUtil.java [hunk 1]
```diff
@@ -95,6 +134,9 @@ public static <T extends Output<T>> MutableDataset<T> generateDataset(DataFrame
                     // TODO: support anomaly labels to evaluate prediction result
                     example = new ArrayExample<>((T) new Event(defaultEventType), featureNamesValues.v1(), featureNamesValues.v2()[i]);
                     break;
+                case LABEL:
+                    example = new ArrayExample<>((T) new Label(""), featureNamesValues.v1(), featureNamesValues.v2()[i]);
```
Comment (Craigacp) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/TribuoUtil.java:138]: You can use [`UNKNOWN_LABEL`](https://tribuo.org/learn/4.2/javadoc/org/tribuo/classification/LabelFactory.html#UNKNOWN_LABEL) for this if you want. It reduces the amount of allocation, and is fine to be shared among examples at prediction time. Or `outputFactory.getUnknownOutput()` as you already have an output factory reference.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/regression/LogisticRegression.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        }
+    }
+
+    @Override
+    public Model train(DataFrame dataFrame) {
+        MutableDataset<Label> trainDataset = TribuoUtil.generateDatasetWithTarget(dataFrame, new LabelFactory(),
+                "Logistic regression training data from OpenSearch", TribuoOutputType.LABEL, parameters.getTarget());
+        // Integer epochs = Optional.ofNullable(parameters.getEpochs()).orElse(DEFAULT_EPOCHS);
+        // LinearSGDTrainer(objective=LogMulticlass,optimiser=AdaGrad(initialLearningRate=1.0,epsilon=0.1,initialValue=0.0),epochs=5,minibatchSize=1,seed=12345)
+        Trainer<Label> logisticRegressionTrainer = new LogisticRegressionTrainer();
```
Comment (Craigacp) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/regression/LogisticRegression.java:119]: Our `LogisticRegressionTrainer` is designed to be very simple for onboarding users & demos, I recommend you use `LinearSGDTrainer` and configure it given you're wrapping all the interface.

Comment (ylwu-amzn) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/regression/LogisticRegression.java:119]: @Craigacp Thanks for your suggestion. We have integrated `LinearSGDTrainer` to provide linear regression https://github.com/opensearch-project/ml-commons/blob/main/ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/regression/LinearRegression.java#L203.  

You mean we should wrap `LinearSGDTrainer` rather than `LogisticRegressionTrainer` to provide logistic regression?

Comment (Craigacp) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/regression/LogisticRegression.java:119]: Well, this is slightly unfortunate naming on our part. We have three classes all called `LinearSGDTrainer`, `org.tribuo.classification.sgd.linear.LinearSGDTrainer`, `org.tribuo.multilabel.sgd.linear.LinearSGDTrainer` and `org.tribuo.regression.sgd.linear.LinearSGDTrainer`. They have basically the same interface, but are constrained by the type system to accept the appropriate kind of objective function, and have the right generic types. `LogisticRegressionTrainer` is just a `org.tribuo.classification.sgd.linear.LinearSGDTrainer` subclass with a constructor that has fairly sensible defaults for running quick tests & demos. So for your use case I'd recommend you use `org.tribuo.classification.sgd.linear.LinearSGDTrainer` as then you can expose basically the same set of options as you do for linear regression (aside from the difference in objective functions).

Comment (ylwu-amzn) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/regression/LogisticRegression.java:119]: Thanks for the details. @Zhangxunmt I think we should change to `org.tribuo.classification.sgd.linear.LinearSGDTrainer`. 

BTW, @Craigacp , For linear regression `org.tribuo.regression.sgd.linear.LinearSGDTrainer` and this `org.tribuo.classification.sgd.linear.LinearSGDTrainer`, so you have any suggestion for the default parameter value like objective function, epochs, batch size etc?

Comment (Craigacp) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/regression/LogisticRegression.java:119]: For the classification/logistic regression the ones we use in the constructor are fine - https://github.com/oracle/tribuo/blob/main/Classification/SGD/src/main/java/org/tribuo/classification/sgd/linear/LogisticRegressionTrainer.java#L44. You could use them for the linear regression too, substituting the `LogMulticlass` objective for `SquaredLoss`. For the linear models a batch size of 1 tends to work just fine, and there isn't really any performance benefit to larger ones given the way the linear algebra system is setup (as it's not multithreaded yet).

Unfortunately we don't have any convergence metrics like validation error built into Tribuo's trainers, and we'd need to modify the trainer API to enable such things.

Comment (Zhangxunmt) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/regression/LogisticRegression.java:119]: I was using LinearSGDTrainer with the default parameters in the first place (check the commented out line), then I realized it's almost equivalent to the given LogisticRegressionTrainer since it's basically using LinearSGDTrainer under the hood. We can change it to LinearSGDTrainer, and let customers to configure parameters through the training requests.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/TribuoUtil.java (hunk 1 above)
Comment (Zhangxunmt) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/TribuoUtil.java]: Replaced with outputFactory.getUnknownOutput().

File: common/src/main/java/org/opensearch/ml/common/dataframe/DefaultDataFrame.java [hunk 3]
```diff
@@ -131,6 +131,24 @@ public DataFrame select(int[] columns) {
         return new DefaultDataFrame(newColumnMetas, rows.stream().map(row-> row.select(columns)).collect(Collectors.toList()));
     }
 
+    @Override
+    public int getColumnIndex(String target) {
+        List<String> featureNames = Arrays.stream(this.columnMetas()).map(ColumnMeta::getName).collect(Collectors.toList());
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/dataframe/DefaultDataFrame.java:136]: Minor: rename `featureNames` as `columnNames`?

File: common/src/main/java/org/opensearch/ml/common/input/parameter/regression/LogisticRegressionParams.java [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+                throw new IllegalArgumentException("Wrong objective type");
+            }
+        }
+    }
+
+    public enum OptimizerType {
+        SIMPLE_SGD,
+        LINEAR_DECAY_SGD,
+        SQRT_DECAY_SGD,
+        ADA_GRAD,
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/input/parameter/regression/LogisticRegressionParams.java:231]: @Craigacp help review if all these supported for logistic regression?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/regression/LogisticRegression.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        if (parameters.getEpochs() != null && parameters.getEpochs() < 0) {
+            throw new IllegalArgumentException("Epochs should not be negative.");
+        }
+
+        if (parameters.getBatchSize() != null && parameters.getBatchSize() < 0) {
+            throw new IllegalArgumentException("MiniBatchSize should not be negative.");
+        }
+
+        if (parameters.getLoggingInterval() != null && parameters.getLoggingInterval() < 0) {
+            throw new IllegalArgumentException("Logging intervals should not be negative.");
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/regression/LogisticRegression.java:91]: From java doc of `org.tribuo.classification.sgd.linear.LinearSGDTrainer`
> Log the loss after this many iterations. If -1 don't log anything.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/regression/LogisticRegression.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        }
+    }
+
+    private void createOptimiser() {
+        LogisticRegressionParams.OptimizerType optimizerType = Optional.ofNullable(parameters.getOptimizerType()).orElse(DEFAULT_OPTIMIZER_TYPE);
+        Double learningRate = Optional.ofNullable(parameters.getLearningRate()).orElse(DEFAULT_LEARNING_RATE);
+        Double epsilon = Optional.ofNullable(parameters.getEpsilon()).orElse(DEFAULT_EPSILON);
+
+        switch (optimizerType) {
+            // ToDo: Add more possible optimizer. Tribuo only provides AdaGrad for logistic regression.
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/regression/LogisticRegression.java:118]: @Craigacp can you confirm if this is true? Does `org.tribuo.classification.sgd.linear.LinearSGDTrainer` support other optimizer ?

File: common/src/main/java/org/opensearch/ml/common/input/parameter/regression/LogisticRegressionParams.java (hunk 4 above)
Comment (Craigacp) [reply common/src/main/java/org/opensearch/ml/common/input/parameter/regression/LogisticRegressionParams.java:231]: These are all supported.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/regression/LogisticRegression.java (hunk 5 above)
Comment (Zhangxunmt) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/regression/LogisticRegression.java:91]: Good catch!

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/regression/LogisticRegression.java (hunk 6 above)
Comment (Zhangxunmt) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/regression/LogisticRegression.java:143]: Thanks for the clarification. These optimizers are all supported in the PR now.

**Review by ylwu-amzn — APPROVED** [review]

**Review by jngz-es — APPROVED** [review]
